---
title: "sar11blastp.Rmd"
output: html_document
date: "2023-04-20"
---

```{r setup, include = FALSE}
library(knitr)
library(tidyverse)
library(kableExtra)
library(DT)
library(reshape2)
library(Biostrings)
library(tm)
```

# Dowload Software

##This is the code to dowload the software. My computer had this software installed already. 

```{bash}
cd /home/joyvan/applications
curl -O https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.13.0+-x64-linux.tar.gz

```

```{bash}
cd /home/joyvan/applications
tar -xf ncbi-blast-2.13.0+-x64-linux.tar.gz
```
```{bash}
~/applications/ncbi-blast-2.13.0+/bin/blastx -h
```

# Make a database

Comparing protein sequence 012276695protein.faa (for easier terms called 95) to 000012345.1protein.faa (for easier terms called 45). First creating a database for 95.  

```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb \
-in ../data_raw/012276695protein.faa \
-dbtype prot \
-out ../blastdb/95-protein
```

Comparing 95 to 45. 
```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/blastp \
-query ../data_raw/000012345.1protein.faa \
-db ../blastdb/95-protein \
-out ../output/95-45_blastp.tab \
-evalue 1E-20 \
-num_threads 20 \
-max_target_seqs 1 \
-outfmt 6
```

Secondly creating a database for 45.
```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb \
-in ../data_raw/000012345.1protein.faa \
-dbtype prot \
-out ../blastdb/45-protein
```

Comparing 45 to 95.
```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/blastp \
-query ../data_raw/012276695protein.faa \
-db ../blastdb/95-protein \
-out ../output/45-95_blastp.tab \
-evalue 1E-20 \
-num_threads 20 \
-max_target_seqs 1 \
-outfmt 6
```


#Get the query sequence

```{bash}
head ../data_raw/012276695protein.faa
echo "How many sequences are there?"
grep -c ">" ../data_raw/012276695protein.faa
```

#Histogram for the Protein Sequences
```{r}
# Read FASTA file
fasta_file <- "../data_raw/012276695protein.faa"  
psequences <- readDNAStringSet(fasta_file)

# Calculate sequence lengths
psequence_lengths <- width(psequences)

# Create a data frame
psequence_lengths_df <- data.frame(Length = psequence_lengths)

# Plot histogram using ggplot2
ggplot(psequence_lengths_df, aes(x = Length)) +
  geom_histogram(binwidth = 1, color = "purple", fill = "red", alpha = 0.75) +
  labs(title = "Histogram of P-Sequence Lengths",
       x = "P-Sequence Length",
       y = "Frequency") +
  theme_minimal()
```

#Boxplots for Query Sequences
```{r}
# Read FASTA file
fasta_file <- "../data_raw/012276695protein.faa"
sequences <- readDNAStringSet(fasta_file)

# Calculate base composition
base_composition <- alphabetFrequency(sequences, baseOnly = TRUE)

# Convert to data frame and reshape for ggplot2
base_composition_df <- as.data.frame(base_composition)
base_composition_df$ID <- rownames(base_composition_df)
base_composition_melted <- reshape2::melt(base_composition_df, id.vars = "ID", variable.name = "Base", value.name = "Count")

# Plot base composition bar chart using ggplot2
ggplot(base_composition_melted, aes(x = Base, y = Count, fill = Base)) +
  geom_bar(stat = "identity", position = "dodge", color = "black") +
  labs(title = "Base Composition",
       x = "Base",
       y = "Count") +
  theme_minimal() +
  scale_fill_manual(values = c("A" = "purple", "C" = "orange", "G" = "pink", "T" = "yellow"))
```


#Run BLAST

```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/blastp \
-query ../data_raw/000012345.1protein.faa \
-db ../blastdb/95-protein \
-out ../output/95-45_blastp.tab \
-evalue 1E-20 \
-num_threads 20 \
-max_target_seqs 1 \
-outfmt 6
```

```{bash}
head -2 ../output/95-45_blastp.tab
wc -l ../output/95-45_blastp.tab
```

#Joining blast table with annotation table

```{bash}
head ../data_raw/012276695protein.faa
echo "How many sequences are there?"
grep -c ">" ../data_raw/012276695protein.faa
```
```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/blastp \
-query ../data_raw/000012345.1protein.faa \
-db ../blastdb/95-protein \
-out ../output/95-45_blastp.tab \
-evalue 1E-20 \
-num_threads 20 \
-max_target_seqs 1 \
-outfmt 6
```

```{bash}
head -2 ../output/95-45_blastp.tab
wc -l ../output/95-45_blastp.tab
```


#Joining blast table with annotation table


```{bash}
head -2 ../output/95-45_blastp.tab
wc -l ../output/95-45_blastp.tab
```


```{bash}
#I replaced this chunk with line 188-190.
#tr '|' '\t' < ../output/95-45_blastp.tab | head -2
```
```{bash}
tr '|' '\t' < ../output/95-45_blastp.tab > temp.txt && head -2 temp.txt
```


```{bash}
tr '|' '\t' < ../output/95-45_blastp.tab \
> ../output/95-45_blastp.tab
```


```{bash}
head -2 ../data_raw/012276695protein.faa
wc -l ../data_raw/012276695protein.faa
```


```{r}
bltabl <- read.csv("../output/95-45_blastp.tab", sep = '\t', header = FALSE)
```

```{r}
spgo <- read.csv("../data_raw/012276695protein.faa", sep = '\t', header = TRUE)
```

```{r}
str(spgo)
```

```{r}
kbl(
head(
  left_join(bltabl, spgo,  by = c("V3" = "Entry")) %>%
  select(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) %>% mutate(V1 = str_replace_all(V1, 
            pattern = "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed", replacement = "Ab"))
)
) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

```{r}
left_join(bltabl, spgo,  by = c("V3" = "Entry")) %>%
  select(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) %>% mutate(V1 = str_replace_all(V1, 
            pattern = "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed", replacement = "Ab")) %>%
write_delim("../output/95_blast_annot_go.tab", delim = '\t')
```


```{r}
annot_tab <- read.csv("../output/blast_annot_go.tab", sep = '\t', header = TRUE)
